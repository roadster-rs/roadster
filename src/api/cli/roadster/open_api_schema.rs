use clap::Parser;
use serde_derive::Serialize;
use std::path::PathBuf;

#[derive(Debug, Parser, Serialize)]
pub struct OpenApiArgs {
    /// The file to write the schema to. If not provided, will write to stdout.
    #[clap(short, long, value_name = "FILE", value_hint = clap::ValueHint::FilePath)]
    pub output: Option<PathBuf>,
    /// Whether to pretty-print the schema. Default: false.
    #[clap(short, long, default_value_t = false)]
    pub pretty_print: bool,
}
