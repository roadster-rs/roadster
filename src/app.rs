use async_trait::async_trait;

use crate::config::app_config::AppConfig;
use crate::context::{AppContext, FromAppContextRef, GetAppContextFromRef};
use crate::tracing::init_tracing;

pub async fn start<A>() -> anyhow::Result<()>
where
    A: App + Default + Sync,
{
    let config = AppConfig::new()?;

    let app = A::default();
    app.init_tracing(&config)?;

    let context = AppContext::new(config).await?;
    let _context = app.context_to_state(&context).await?;

    Ok(())
}

#[async_trait]
pub trait App {
    type State: FromAppContextRef + GetAppContextFromRef;

    fn init_tracing(&self, config: &AppConfig) -> anyhow::Result<()> {
        init_tracing(config)?;

        Ok(())
    }

    async fn context_to_state(&self, app_context: &AppContext) -> anyhow::Result<Self::State> {
        let state = Self::State::from_app_context_ref(app_context).await?;
        Ok(*state)
    }
}
