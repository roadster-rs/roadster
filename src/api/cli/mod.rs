use crate::api::cli::roadster::{RoadsterCli, RunRoadsterCommand};
use crate::app::context::AppContext;
use crate::app::App;
#[cfg(test)]
use crate::app::MockApp;
use crate::error::RoadsterResult;
use crate::migration::Migrator;
use crate::service::registry::ServiceRegistry;
use async_trait::async_trait;
use axum_core::extract::FromRef;
use clap::{Args, Command, FromArgMatches};
use std::ffi::OsString;

pub mod roadster;

#[non_exhaustive]
pub struct CliState<A, S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S> + Sync + 'static,
{
    pub roadster_cli: RoadsterCli,
    pub app_cli: A::Cli,
    pub app: A,
    pub state: S,
    #[cfg(feature = "db-sql")]
    pub migrators: Vec<Box<dyn Migrator<S>>>,
    pub service_registry: ServiceRegistry<A, S>,
}

/// Implement to enable Roadster to run your custom CLI commands.
#[async_trait]
pub trait RunCommand<A, S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S> + Sync,
{
    /// Run the command.
    ///
    /// # Returns
    /// * `Ok(true)` - If the implementation handled the command and thus the app should end execution
    ///     after the command is complete.
    /// * `Ok(false)` - If the implementation did not handle the command and thus the app should
    ///     continue execution after the command is complete.
    /// * `Err(...)` - If the implementation experienced an error while handling the command. The
    ///     app should end execution after the command is complete.
    async fn run(&self, cli: &CliState<A, S>) -> RoadsterResult<bool>;
}

pub(crate) fn parse_cli<A, S, I, T>(args: I) -> RoadsterResult<(RoadsterCli, A::Cli)>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
    I: IntoIterator<Item = T>,
    T: Into<OsString> + Clone,
{
    // Build the CLI by augmenting a default Command with both the roadster and app-specific CLIs
    let cli = Command::default();
    // Add the roadster CLI. Save the shared attributes to use after adding the app-specific CLI
    let cli = RoadsterCli::augment_args(cli);
    let about = cli.get_about().cloned();
    let long_about = cli.get_long_about().cloned();
    let version = cli.get_version().map(|x| x.to_string());
    let long_version = cli.get_long_version().map(|x| x.to_string());
    // Add the app-specific CLI. This will override the shared attributes, so we need to
    // combine them with the roadster CLI attributes.
    let cli = A::Cli::augment_args(cli);
    let cli = if let Some((a, b)) = about.zip(cli.get_about().cloned()) {
        cli.about(format!("{a}\n\n{b}"))
    } else {
        cli
    };
    let cli = if let Some((a, b)) = long_about.zip(cli.get_long_about().cloned()) {
        cli.long_about(format!("{a}\n\n{b}"))
    } else {
        cli
    };
    let cli = if let Some((a, b)) = version.zip(cli.get_version().map(|x| x.to_string())) {
        cli.version(format!("roadster: {a}, app: {b}"))
    } else {
        cli
    };
    let cli = if let Some((a, b)) = long_version.zip(cli.get_long_version().map(|x| x.to_string()))
    {
        cli.long_version(format!("roadster: {a}\n\napp: {b}"))
    } else {
        cli
    };
    // Build each CLI from the CLI args
    let matches = cli.get_matches_from(args);
    let roadster_cli = RoadsterCli::from_arg_matches(&matches)?;
    let app_cli = A::Cli::from_arg_matches(&matches)?;
    Ok((roadster_cli, app_cli))
}

pub(crate) async fn handle_cli<A, S>(cli: &CliState<A, S>) -> RoadsterResult<bool>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    if cli.roadster_cli.run(cli).await? {
        return Ok(true);
    }
    if cli.app_cli.run(cli).await? {
        return Ok(true);
    }
    Ok(false)
}

#[cfg(test)]
pub struct TestCli<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    _state: std::marker::PhantomData<S>,
}

#[cfg(test)]
mockall::mock! {
    pub TestCli<S>
    where
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
    {}

    #[async_trait]
    impl<S> RunCommand<MockApp<S>, S> for TestCli<S>
    where
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
    {
        async fn run(&self, prepared: &CliState<MockApp<S>, S>) -> RoadsterResult<bool>;
    }

    impl<S> clap::FromArgMatches for TestCli<S>
    where
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
    {
        fn from_arg_matches(matches: &clap::ArgMatches) -> Result<Self, clap::Error>;
        fn update_from_arg_matches(&mut self, matches: &clap::ArgMatches) -> Result<(), clap::Error>;
    }

    impl<S> clap::Args for TestCli<S>
    where
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
    {
        fn augment_args(cmd: clap::Command) -> clap::Command;
        fn augment_args_for_update(cmd: clap::Command) -> clap::Command;
    }

    impl<S> Clone for TestCli<S>
    where
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
    {
        fn clone(&self) -> Self;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::app::MockApp;
    use crate::service::registry::ServiceRegistry;
    use crate::testing::snapshot::TestCase;
    use insta::assert_toml_snapshot;
    use itertools::Itertools;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[cfg_attr(coverage_nightly, coverage(off))]
    fn setup_cli(
        args: Vec<&str>,
        mock_handles_cli: bool,
    ) -> (RoadsterCli, MockTestCli<AppContext>) {
        let augment_args_context = MockTestCli::<AppContext>::augment_args_context();
        augment_args_context.expect().returning(|c| c);
        let from_arg_matches_context = MockTestCli::<AppContext>::from_arg_matches_context();
        from_arg_matches_context
            .expect()
            .returning(|_| Ok(MockTestCli::<AppContext>::default()));

        let mut app_cli = MockTestCli::<AppContext>::default();
        app_cli
            .expect_run()
            .returning(move |_| Ok(mock_handles_cli));

        // The first word is interpreted as the binary name
        let args = vec!["binary_name"].into_iter().chain(args).collect_vec();

        let (roadster_cli, _) = super::parse_cli::<MockApp<AppContext>, _, _, _>(args).unwrap();

        (roadster_cli, app_cli)
    }

    #[rstest]
    #[case(None)]
    #[case(Some("--environment test"))]
    #[case(Some("--skip-validate-config"))]
    #[case(Some("--allow-dangerous"))]
    #[cfg_attr(feature = "open-api", case::list_routes(Some("roadster list-routes")))]
    #[cfg_attr(feature = "open-api", case::list_routes(Some("r list-routes")))]
    #[cfg_attr(feature = "open-api", case::open_api(Some("r open-api")))]
    #[cfg_attr(feature = "db-sql", case::migrate(Some("r migrate up")))]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn parse_cli(_case: TestCase, #[case] args: Option<&str>) {
        // Arrange
        let args = args
            .map(|args| args.split(' ').collect_vec())
            .unwrap_or_default();

        // Act
        let (roadster_cli, _a) = setup_cli(args, false);

        // Assert
        assert_toml_snapshot!(roadster_cli);
    }

    #[rstest]
    #[case(None, false, false)]
    #[case(None, true, true)]
    #[cfg_attr(
        feature = "open-api",
        case::list_routes(Some("roadster handle-cli"), false, true)
    )]
    #[tokio::test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    async fn handle_cli(
        _case: TestCase,
        #[case] args: Option<&str>,
        #[case] mock_handles_cli: bool,
        #[case] cli_handled: bool,
    ) {
        let context = AppContext::test(None, None, None).unwrap();
        let app = MockApp::default();

        let args = args
            .map(|args| args.split(' ').collect_vec())
            .unwrap_or_default();

        let (roadster_cli, app_cli) = setup_cli(args, mock_handles_cli);

        let cli = CliState {
            roadster_cli,
            app_cli,
            app,
            #[cfg(feature = "db-sql")]
            migrators: Default::default(),
            service_registry: ServiceRegistry::new(&context),
            state: context,
        };

        let result = super::handle_cli(&cli).await.unwrap();

        assert_eq!(result, cli_handled);
    }
}
