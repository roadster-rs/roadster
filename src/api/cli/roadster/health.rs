use crate::api::cli::roadster::RunRoadsterCommand;
use crate::api::cli::CliState;
use crate::api::core::health::health_check;
use crate::app::context::AppContext;
use crate::app::App;
use crate::error::RoadsterResult;
use async_trait::async_trait;
use axum_core::extract::FromRef;
use clap::Parser;
use serde_derive::Serialize;
use std::time::Duration;
use tracing::info;

#[derive(Debug, Parser, Serialize)]
#[non_exhaustive]
pub struct HealthArgs {
    /// Maximum time to spend checking the health of the resources in milliseconds
    #[clap(short = 'd', long)]
    max_duration: Option<u64>,
}

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for HealthArgs
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, cli: &CliState<A, S>) -> RoadsterResult<bool> {
        let duration = self
            .max_duration
            .map(Duration::from_millis)
            .unwrap_or_else(|| {
                let context = AppContext::from_ref(&cli.state);
                context.config().health_check.max_duration.cli
            });
        let health = health_check(&cli.state, Some(duration)).await?;
        let health = serde_json::to_string_pretty(&health)?;
        info!("\n{health}");
        Ok(true)
    }
}
