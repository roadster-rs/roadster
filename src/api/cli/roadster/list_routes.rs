use crate::api::cli::roadster::RunRoadsterCommand;
use crate::app::context::AppContext;
use crate::app::{App, PreparedApp};
use crate::error::RoadsterResult;
use crate::service::http::service::HttpService;
use anyhow::anyhow;
use async_trait::async_trait;
use axum_core::extract::FromRef;
use clap::Parser;
use itertools::Itertools;
use serde_derive::Serialize;
use tracing::info;

#[derive(Debug, Parser, Serialize)]
#[non_exhaustive]
pub struct ListRoutesArgs {}

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for ListRoutesArgs
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, prepared_app: &PreparedApp<A, S>) -> RoadsterResult<bool> {
        let http_service = prepared_app
            .service_registry
            .get::<HttpService>()
            .map_err(|err| {
                anyhow!("Unable to get HttpService from registry. Was it registered? Err: {err}")
            })?;

        let routes = http_service
            .list_routes()
            .into_iter()
            .map(|(path, method)| format!("[{method}]\t{path}"))
            .join("\n\t");

        info!("API routes:\n\t{routes}");

        Ok(true)
    }
}
