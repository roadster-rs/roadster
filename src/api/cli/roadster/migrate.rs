use crate::api::cli::CliState;
use crate::api::cli::roadster::RunRoadsterCommand;
use crate::app::App;
use crate::app::context::AppContext;
use crate::db::migration::{DownArgs, MigrationInfo, UpArgs};
use crate::error::RoadsterResult;
use async_trait::async_trait;
use axum_core::extract::FromRef;
use clap::{Parser, Subcommand};
use itertools::Itertools;
use serde_derive::Serialize;
use tracing::{info, warn};

#[derive(Debug, Parser, Serialize)]
#[non_exhaustive]
pub struct MigrateArgs {
    #[clap(subcommand)]
    pub command: MigrateCommand,
}

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for MigrateArgs
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, cli: &CliState<A, S>) -> RoadsterResult<bool> {
        self.command.run(cli).await
    }
}

#[derive(Debug, Subcommand, Serialize)]
#[serde(tag = "type")]
#[non_exhaustive]
pub enum MigrateCommand {
    /// Apply pending migrations. If no `steps` argument is provided, will apply all pending
    /// migrations.
    Up(UpArgs),
    /// Roll back applied migrations. If no `steps` argument is provided, will roll back all
    /// applied migrations.
    Down(DownArgs),
    /// Check the status of all migrations
    Status,
}

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for MigrateCommand
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, cli: &CliState<A, S>) -> RoadsterResult<bool> {
        let context = AppContext::from_ref(&cli.state);

        let allow_dangerous = cli.roadster_cli.allow_dangerous(&context);

        if is_destructive(self) && !allow_dangerous {
            return Err(crate::error::cli::CliError::DestructiveCmdNotAllowed(
                context.config().environment.clone(),
            )
            .into());
        } else if is_destructive(self) {
            warn!(
                "Running destructive command `{:?}` in environment `{:?}`",
                self,
                context.config().environment
            );
        }

        match self {
            MigrateCommand::Up(args) => {
                migrate_up(cli, args).await?;
                print_status(cli).await?;
            }
            MigrateCommand::Down(args) => {
                migrate_down(cli, args).await?;
                print_status(cli).await?;
            }
            MigrateCommand::Status => {
                print_status(cli).await?;
            }
        };
        Ok(true)
    }
}

fn is_destructive(command: &MigrateCommand) -> bool {
    !matches!(command, MigrateCommand::Status)
}

// Todo: reduce duplication
async fn migrate_up<A, S>(cli: &CliState<A, S>, args: &UpArgs) -> RoadsterResult<()>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    let mut total_steps_run = 0;
    for migrator in cli.migrators.iter() {
        let remaining_steps = args
            .steps
            .map(|steps| steps.saturating_sub(total_steps_run));
        if let Some(remaining) = remaining_steps {
            if remaining == 0 {
                return Ok(());
            }
        }
        let steps_run = migrator
            .up(
                &cli.state,
                &UpArgs::builder().maybe_steps(remaining_steps).build(),
            )
            .await?;
        total_steps_run += steps_run;
    }
    Ok(())
}

// Todo: reduce duplication
async fn migrate_down<A, S>(cli: &CliState<A, S>, args: &DownArgs) -> RoadsterResult<()>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    let mut total_steps_run = 0;
    for migrator in cli.migrators.iter().rev() {
        let remaining_steps = args
            .steps
            .map(|steps| steps.saturating_sub(total_steps_run));
        if let Some(remaining) = remaining_steps {
            if remaining == 0 {
                return Ok(());
            }
        }
        let steps_run = migrator
            .down(
                &cli.state,
                &DownArgs::builder().maybe_steps(remaining_steps).build(),
            )
            .await?;
        total_steps_run += steps_run;
    }
    Ok(())
}

async fn print_status<A, S>(cli: &CliState<A, S>) -> RoadsterResult<()>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    let mut migrations: Vec<MigrationInfo> = Vec::new();
    for migrator in cli.migrators.iter() {
        migrations.extend(migrator.status(&cli.state).await?);
    }
    let migrations = migrations
        .into_iter()
        .map(|migration| {
            let status: &'static str = migration.status.into();
            format!("{}\t{}", status, migration.name)
        })
        .join("\n");
    info!("Migration status:\n{migrations}");
    Ok(())
}
