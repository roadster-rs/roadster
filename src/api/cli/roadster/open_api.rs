use crate::api::cli::roadster::RunRoadsterCommand;
use crate::app::context::AppContext;
use crate::app::{App, PreparedApp};
use crate::error::RoadsterResult;
use crate::service::http::service::{HttpService, OpenApiArgs};
use anyhow::anyhow;
use async_trait::async_trait;
use axum_core::extract::FromRef;

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for OpenApiArgs
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, prepared_app: &PreparedApp<A, S>) -> RoadsterResult<bool> {
        let http_service = prepared_app
            .service_registry
            .get::<HttpService>()
            .map_err(|err| {
                anyhow!("Unable to get HttpService from registry. Was it registered? Err: {err}")
            })?;

        http_service.print_open_api_schema(self)?;

        Ok(true)
    }
}
