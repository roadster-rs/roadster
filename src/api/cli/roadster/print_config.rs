use async_trait::async_trait;
use axum_core::extract::FromRef;
use clap::Parser;
use serde_derive::{Deserialize, Serialize};
use strum_macros::{EnumString, IntoStaticStr};
use tracing::info;

use crate::api::cli::roadster::{RoadsterCli, RunRoadsterCommand};
use crate::app::context::AppContext;
use crate::app::App;
use crate::config::AppConfig;
use crate::error::RoadsterResult;

#[derive(Debug, Parser, Serialize)]
#[non_exhaustive]
pub struct PrintConfigArgs {
    /// Print the config with the specified format.
    #[clap(short, long, default_value = "debug")]
    pub format: Format,
}

#[derive(
    Debug, Clone, Eq, PartialEq, Serialize, Deserialize, EnumString, IntoStaticStr, clap::ValueEnum,
)]
#[serde(rename_all = "kebab-case", tag = "type")]
#[strum(serialize_all = "kebab-case")]
#[non_exhaustive]
pub enum Format {
    Debug,
    Json,
    JsonPretty,
    Toml,
    TomlPretty,
}

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for PrintConfigArgs
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, _app: &A, _cli: &RoadsterCli, state: &S) -> RoadsterResult<bool> {
        let context = AppContext::from_ref(state);
        let serialized = serialize_config(&self.format, context.config())?;

        info!("\n{}", serialized);

        Ok(true)
    }
}

fn serialize_config(format: &Format, config: &AppConfig) -> RoadsterResult<String> {
    let serialized = match format {
        Format::Debug => {
            format!("{:?}", config)
        }
        Format::Json => serde_json::to_string(config)?,
        Format::JsonPretty => serde_json::to_string_pretty(config)?,
        Format::Toml => toml::to_string(config)?,
        Format::TomlPretty => toml::to_string_pretty(config)?,
    };
    Ok(serialized)
}

#[cfg(all(
    test,
    feature = "http",
    feature = "open-api",
    feature = "sidekiq",
    feature = "db-sql",
    feature = "email-smtp",
    feature = "email-sendgrid",
    feature = "jwt",
    feature = "jwt-ietf",
    feature = "otel",
    feature = "grpc",
    feature = "testing",
    feature = "test-containers",
    feature = "testing-mocks",
    feature = "config-yml",
))]
mod tests {
    use crate::api::cli::roadster::print_config::Format;
    use crate::config::AppConfig;
    use crate::testing::snapshot::TestCase;
    use insta::assert_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn config() -> AppConfig {
        AppConfig::test(None).unwrap()
    }

    #[rstest]
    #[case(Format::Debug)]
    #[case(Format::Json)]
    #[case(Format::JsonPretty)]
    #[case(Format::Toml)]
    #[case(Format::TomlPretty)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn serialize_config(_case: TestCase, config: AppConfig, #[case] format: Format) {
        let serialized = super::serialize_config(&format, &config).unwrap();

        assert_snapshot!(serialized);
    }
}
