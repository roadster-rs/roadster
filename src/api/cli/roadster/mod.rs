use crate::api::cli::roadster::health::HealthArgs;
#[cfg(feature = "open-api")]
use crate::api::cli::roadster::list_routes::ListRoutesArgs;
#[cfg(feature = "db-sea-orm")]
use crate::api::cli::roadster::migrate::MigrateArgs;
use crate::api::cli::roadster::print_config::PrintConfigArgs;
use crate::app::context::AppContext;
use crate::app::App;
use crate::config::environment::Environment;
use crate::error::RoadsterResult;
#[cfg(feature = "open-api")]
use crate::service::http::service::OpenApiArgs;
use async_trait::async_trait;
use std::path::PathBuf;

use axum_core::extract::FromRef;
use clap::{Parser, Subcommand};
use serde_derive::Serialize;

pub mod health;
#[cfg(feature = "open-api")]
pub mod list_routes;
#[cfg(feature = "db-sea-orm")]
pub mod migrate;
pub mod print_config;

/// Internal version of [RunCommand][crate::cli::RunCommand] that uses the [RoadsterCli] and
/// [AppContext] instead of the consuming app's versions of these objects. This (slightly) reduces
/// the boilerplate required to implement a Roadster command.
#[async_trait]
pub(crate) trait RunRoadsterCommand<A, S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, app: &A, cli: &RoadsterCli, state: &S) -> RoadsterResult<bool>;
}

/// Roadster: The Roadster CLI provides various utilities for managing your application. If no subcommand
/// is matched, Roadster will default to running/serving your application.
#[derive(Debug, Parser, Serialize)]
#[command(version, about)]
#[non_exhaustive]
pub struct RoadsterCli {
    /// Specify the environment to use to run the application. This overrides the corresponding
    /// environment variable if it's set.
    #[clap(short, long)]
    pub environment: Option<Environment>,

    /// Skip validation of the app config. This can be useful for debugging the app config
    /// when used in conjunction with the `print-config` command.
    #[clap(long, action)]
    pub skip_validate_config: bool,

    /// Allow dangerous/destructive operations when running in the `production` environment. If
    /// this argument is not provided, dangerous/destructive operations will not be performed
    /// when running in `production`.
    #[clap(long, action)]
    pub allow_dangerous: bool,

    /// The location of the config directory (where the app's config files are located). If
    /// not provided, will default to `./config/`.
    #[clap(long, value_name = "CONFIG_DIRECTORY", value_hint = clap::ValueHint::DirPath)]
    pub config_dir: Option<PathBuf>,

    #[command(subcommand)]
    pub command: Option<RoadsterCommand>,
}

impl RoadsterCli {
    pub fn allow_dangerous(&self, context: &AppContext) -> bool {
        context.config().environment != Environment::Production || self.allow_dangerous
    }
}

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for RoadsterCli
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, app: &A, cli: &RoadsterCli, state: &S) -> RoadsterResult<bool> {
        if let Some(command) = self.command.as_ref() {
            command.run(app, cli, state).await
        } else {
            Ok(false)
        }
    }
}

#[derive(Debug, Subcommand, Serialize)]
#[serde(tag = "type")]
#[non_exhaustive]
pub enum RoadsterCommand {
    /// Roadster subcommands. Subcommands provided by Roadster are listed under this subcommand in
    /// order to avoid naming conflicts with the consumer's subcommands.
    #[clap(visible_alias = "r")]
    Roadster(RoadsterArgs),
}

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for RoadsterCommand
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, app: &A, cli: &RoadsterCli, state: &S) -> RoadsterResult<bool> {
        match self {
            RoadsterCommand::Roadster(args) => args.run(app, cli, state).await,
        }
    }
}

#[derive(Debug, Parser, Serialize)]
#[non_exhaustive]
pub struct RoadsterArgs {
    #[command(subcommand)]
    pub command: RoadsterSubCommand,
}

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for RoadsterArgs
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, app: &A, cli: &RoadsterCli, state: &S) -> RoadsterResult<bool> {
        self.command.run(app, cli, state).await
    }
}

#[async_trait]
impl<A, S> RunRoadsterCommand<A, S> for RoadsterSubCommand
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    A: App<S>,
{
    async fn run(&self, app: &A, cli: &RoadsterCli, state: &S) -> RoadsterResult<bool> {
        match self {
            #[cfg(feature = "open-api")]
            RoadsterSubCommand::ListRoutes(_) => {
                #[allow(unused_doc_comments)]
                /// Implemented by [crate::service::http::service::HttpService]
                Ok(false)
            }
            #[cfg(feature = "open-api")]
            RoadsterSubCommand::OpenApi(_) => {
                #[allow(unused_doc_comments)]
                /// Implemented by [crate::service::http::service::HttpService]
                Ok(false)
            }
            #[cfg(feature = "db-sea-orm")]
            RoadsterSubCommand::Migrate(args) => args.run(app, cli, state).await,
            RoadsterSubCommand::PrintConfig(args) => args.run(app, cli, state).await,
            RoadsterSubCommand::Health(args) => args.run(app, cli, state).await,
            #[cfg(test)]
            RoadsterSubCommand::HandleCli => Ok(true),
        }
    }
}

#[derive(Debug, Subcommand, Serialize)]
#[serde(tag = "type")]
#[non_exhaustive]
pub enum RoadsterSubCommand {
    /// List the API routes available in the app. Note: only the routes defined
    /// using the `Aide` crate will be included in the output.
    #[cfg(feature = "open-api")]
    ListRoutes(ListRoutesArgs),

    /// Generate an OpenAPI 3.1 schema for the app's API routes. Note: only the routes defined
    /// using the `Aide` crate will be included in the schema.
    #[cfg(feature = "open-api")]
    OpenApi(OpenApiArgs),

    /// Perform DB operations using SeaORM migrations.
    #[cfg(feature = "db-sea-orm")]
    #[clap(visible_aliases = ["m", "migration"])]
    Migrate(MigrateArgs),

    /// Print the AppConfig
    PrintConfig(PrintConfigArgs),

    /// Check the health of the app's resources. Note: This runs without starting the app's service(s)
    /// and only requires creating the [AppContext] that would normally be used by the app.
    Health(HealthArgs),

    /// A test-only command to test handling the CLI.
    #[cfg(test)]
    HandleCli,
}
