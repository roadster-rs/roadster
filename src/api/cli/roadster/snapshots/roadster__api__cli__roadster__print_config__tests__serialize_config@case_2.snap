---
source: src/api/cli/roadster/print_config.rs
expression: serialized
---
{"environment":"test","app":{"name":"Test","shutdown-on-error":true},"lifecycle-handler":{"default-enable":true,"db-migration":{"priority":0},"db-graceful-shutdown":{"priority":10000}},"health-check":{"default-enable":true,"max-duration":{"startup":60000,"api":1000,"cli":10000},"database":{},"sidekiq":{},"smtp":{}},"service":{"default-enable":true,"http":{"url":null,"scheme":"http","host":"127.0.0.1","port":3000,"middleware":{"default-enable":true,"sensitive-request-headers":{"priority":-10000,"header-names":["authorization","refresh-token","x-refresh-token","proxy-authorization","cookie","set-cookie"]},"sensitive-response-headers":{"priority":10000,"header-names":["authorization","refresh-token","x-refresh-token","proxy-authorization","cookie","set-cookie"]},"set-request-id":{"priority":-9990,"header-name":"request-id"},"propagate-request-id":{"priority":9990,"header-name":"request-id"},"tracing":{"priority":-9980},"catch-panic":{"priority":0},"response-compression":{"priority":0},"request-decompression":{"priority":-9960},"timeout":{"priority":0,"timeout":10000},"size-limit":{"priority":-9970,"limit":"5 MB"},"cors":{"priority":-9950,"preset":"restrictive","max-age":3600000},"request-response-logging":{"priority":0,"max-len":1000},"cache-control":{"priority":0,"max-age":604800,"content-types":{}},"etag":{"priority":0}},"initializer":{"default-enable":true,"normalize-path":{"priority":10000}},"default-routes":{"default-enable":true,"ping":{"route":"_ping"},"health":{"route":"_health"},"api-schema":{"route":"_docs/api.json"},"scalar":{"route":"_docs"},"redoc":{"route":"_docs/redoc"}}},"grpc":{"scheme":"http","host":"127.0.0.1","port":3001},"sidekiq":{"num-workers":16,"balance-strategy":"round-robin","queues":[],"redis":{"uri":"redis://[Sensitive]","enqueue-pool":{"min-idle":null,"max-connections":null},"fetch-pool":{"min-idle":null,"max-connections":null},"test-container":null},"periodic":{"stale-cleanup":"auto-clean-stale"},"app-worker":{"max-retries":25,"timeout":true,"max-duration":60},"queue-config":{}}},"auth":{"jwt":{"cookie-name":null,"secret":"secret-test","claims":{"audience":[],"required-claims":[]}}},"tracing":{"level":"debug","format":"compact","service-name":null,"trace-propagation":true,"trace-sampling-ratio":null,"metrics-export-interval":null,"trace-filters":["h2=warn","tower::buffer::worker=warn","opentelemetry_sdk=warn"],"otlp":null},"database":{"uri":"postgres://[Sensitive]","auto-migrate":true,"connect-timeout":1000,"connect-lazy":true,"acquire-timeout":1000,"idle-timeout":null,"max-lifetime":null,"min-connections":0,"max-connections":10,"test-on-checkout":true,"retry-connection":true,"temporary-test-db":false,"temporary-test-db-clean-up":true,"test-container":null},"email":{"from":"no-reply@example.com","reply-to":null,"smtp":{"connection":{"uri":"smtps://username:password@smtp.example.com:425"},"pool":null},"sendgrid":{"api-key":"api-key","sandbox":true,"https-only":true}}}
